// @web3auth/torus-evm-adapter/dist/torusEvmAdapter.esm.js
import _asyncToGenerator from '@babel/runtime/helpers/asyncToGenerator';
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _assertThisInitialized from '@babel/runtime/helpers/assertThisInitialized';
import _get from '@babel/runtime/helpers/get';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _regeneratorRuntime from '@babel/runtime/regenerator';
import Torus from '@toruslabs/torus-embed';
import { WALLET_ADAPTERS, ADAPTER_NAMESPACES, CHAIN_NAMESPACES, ADAPTER_CATEGORY, ADAPTER_STATUS, log, ADAPTER_EVENTS, getChainConfig, Web3AuthError, WalletLoginError, WalletInitializationError } from '@web3auth/base';
import { BaseEvmAdapter } from '@web3auth/base-evm-adapter';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var TorusWalletAdapter = /*#__PURE__*/function (_BaseEvmAdapter) {
  _inherits(TorusWalletAdapter, _BaseEvmAdapter);

  var _super = _createSuper(TorusWalletAdapter);

  function TorusWalletAdapter(params) {
    var _this;

    _classCallCheck(this, TorusWalletAdapter);

    _this = _super.call(this, params);

    _defineProperty(_assertThisInitialized(_this), "name", WALLET_ADAPTERS.TORUS_EVM);

    _defineProperty(_assertThisInitialized(_this), "adapterNamespace", ADAPTER_NAMESPACES.EIP155);

    _defineProperty(_assertThisInitialized(_this), "currentChainNamespace", CHAIN_NAMESPACES.EIP155);

    _defineProperty(_assertThisInitialized(_this), "type", ADAPTER_CATEGORY.EXTERNAL);

    _defineProperty(_assertThisInitialized(_this), "status", ADAPTER_STATUS.NOT_READY);

    _defineProperty(_assertThisInitialized(_this), "torusInstance", null);

    _defineProperty(_assertThisInitialized(_this), "torusWalletOptions", void 0);

    _defineProperty(_assertThisInitialized(_this), "initParams", void 0);

    _defineProperty(_assertThisInitialized(_this), "loginSettings", {});

    _defineProperty(_assertThisInitialized(_this), "rehydrated", false);

    _this.torusWalletOptions = params.adapterSettings || {};
    _this.initParams = params.initParams || {};
    _this.loginSettings = params.loginSettings || {};
    _this.chainConfig = params.chainConfig || null;
    _this.sessionTime = params.sessionTime || 86400;
    return _this;
  }

  _createClass(TorusWalletAdapter, [{
    key: "provider",
    get: function get() {
      if (this.status === ADAPTER_STATUS.CONNECTED && this.torusInstance) {
        return this.torusInstance.provider;
      }

      return null;
    },
    set: function set(_) {
      throw new Error("Not implemented");
    }
  }, {
    key: "init",
    value: function () {
      var _init = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(options) {
        var network, _this$chainConfig, blockExplorer, displayName, chainId, ticker, tickerName, _this$chainConfig2, _chainId2, _blockExplorer, _displayName, rpcTarget, _ticker, _tickerName;

        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _get(_getPrototypeOf(TorusWalletAdapter.prototype), "checkInitializationRequirements", this).call(this); // set chainConfig for mainnet by default if not set


                if (!this.chainConfig) {
                  this.chainConfig = getChainConfig(CHAIN_NAMESPACES.EIP155, 1);
                  _this$chainConfig = this.chainConfig, blockExplorer = _this$chainConfig.blockExplorer, displayName = _this$chainConfig.displayName, chainId = _this$chainConfig.chainId, ticker = _this$chainConfig.ticker, tickerName = _this$chainConfig.tickerName;
                  network = {
                    chainId: Number.parseInt(chainId, 16),
                    host: "mainnet",
                    blockExplorer: blockExplorer,
                    networkName: displayName,
                    ticker: ticker,
                    tickerName: tickerName
                  };
                } else {
                  _this$chainConfig2 = this.chainConfig, _chainId2 = _this$chainConfig2.chainId, _blockExplorer = _this$chainConfig2.blockExplorer, _displayName = _this$chainConfig2.displayName, rpcTarget = _this$chainConfig2.rpcTarget, _ticker = _this$chainConfig2.ticker, _tickerName = _this$chainConfig2.tickerName;
                  network = {
                    chainId: Number.parseInt(_chainId2, 16),
                    host: rpcTarget,
                    blockExplorer: _blockExplorer,
                    networkName: _displayName,
                    ticker: _ticker,
                    tickerName: _tickerName
                  };
                }

                this.torusInstance = new Torus(this.torusWalletOptions);
                log.debug("initializing torus evm adapter init");
                _context.next = 6;
                return this.torusInstance.init(_objectSpread(_objectSpread({
                  showTorusButton: false
                }, this.initParams), {}, {
                  network: network
                }));

              case 6:
                this.status = ADAPTER_STATUS.READY;
                this.emit(ADAPTER_EVENTS.READY, WALLET_ADAPTERS.TORUS_EVM);
                _context.prev = 8;
                log.debug("initializing torus evm adapter");

                if (!options.autoConnect) {
                  _context.next = 14;
                  break;
                }

                this.rehydrated = true;
                _context.next = 14;
                return this.connect();

              case 14:
                _context.next = 20;
                break;

              case 16:
                _context.prev = 16;
                _context.t0 = _context["catch"](8);
                log.error("Failed to connect with torus evm provider", _context.t0);
                this.emit(ADAPTER_EVENTS.ERRORED, _context.t0);

              case 20:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[8, 16]]);
      }));

      function init(_x) {
        return _init.apply(this, arguments);
      }

      return init;
    }()
  }, {
    key: "connect",
    value: function () {
      var _connect = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
        var chainId, _this$chainConfig3, _chainId, blockExplorer, displayName, rpcTarget, ticker, tickerName, network, updatedChainID;

        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _get(_getPrototypeOf(TorusWalletAdapter.prototype), "checkConnectionRequirements", this).call(this);

                if (this.torusInstance) {
                  _context2.next = 3;
                  break;
                }

                throw WalletInitializationError.notReady("Torus wallet is not initialized");

              case 3:
                this.status = ADAPTER_STATUS.CONNECTING;
                this.emit(ADAPTER_EVENTS.CONNECTING, {
                  adapter: WALLET_ADAPTERS.TORUS_EVM
                });
                _context2.prev = 5;
                _context2.next = 8;
                return this.torusInstance.login(this.loginSettings);

              case 8:
                chainId = this.torusInstance.provider.chainId;

                if (!(chainId && parseInt(chainId) !== parseInt(this.chainConfig.chainId, 16))) {
                  _context2.next = 19;
                  break;
                }

                _this$chainConfig3 = this.chainConfig, _chainId = _this$chainConfig3.chainId, blockExplorer = _this$chainConfig3.blockExplorer, displayName = _this$chainConfig3.displayName, rpcTarget = _this$chainConfig3.rpcTarget, ticker = _this$chainConfig3.ticker, tickerName = _this$chainConfig3.tickerName;
                network = {
                  chainId: Number.parseInt(_chainId, 16),
                  host: rpcTarget,
                  blockExplorer: blockExplorer,
                  networkName: displayName,
                  tickerName: tickerName,
                  ticker: ticker
                }; // in some cases when user manually switches chain and relogin then adapter will not connect to initially passed
                // chainConfig but will connect to the one that user switched to.
                // So here trying to switch network to the one that was initially passed in chainConfig.

                _context2.next = 14;
                return this.torusInstance.setProvider(_objectSpread({}, network));

              case 14:
                _context2.next = 16;
                return this.torusInstance.ethereum.request({
                  method: "eth_chainId"
                });

              case 16:
                updatedChainID = _context2.sent;

                if (!(updatedChainID && parseInt(updatedChainID) !== parseInt(this.chainConfig.chainId, 16))) {
                  _context2.next = 19;
                  break;
                }

                throw WalletInitializationError.fromCode(5000, "Not connected to correct chainId. Expected: ".concat(this.chainConfig.chainId, ", Current: ").concat(updatedChainID));

              case 19:
                this.status = ADAPTER_STATUS.CONNECTED;
                this.torusInstance.showTorusButton();
                this.emit(ADAPTER_STATUS.CONNECTED, {
                  adapter: WALLET_ADAPTERS.TORUS_EVM,
                  reconnected: this.rehydrated
                });
                return _context2.abrupt("return", this.provider);

              case 25:
                _context2.prev = 25;
                _context2.t0 = _context2["catch"](5);
                // ready again to be connected
                this.status = ADAPTER_STATUS.READY;
                this.rehydrated = false;
                this.emit(ADAPTER_STATUS.ERRORED, _context2.t0);
                throw _context2.t0 instanceof Web3AuthError ? _context2.t0 : WalletLoginError.connectionError("Failed to login with torus wallet");

              case 31:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this, [[5, 25]]);
      }));

      function connect() {
        return _connect.apply(this, arguments);
      }

      return connect;
    }()
  }, {
    key: "disconnect",
    value: function () {
      var _disconnect = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee3() {
        var options,
            _args3 = arguments;
        return _regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                options = _args3.length > 0 && _args3[0] !== undefined ? _args3[0] : {
                  cleanup: false
                };

                if (this.torusInstance) {
                  _context3.next = 3;
                  break;
                }

                throw WalletInitializationError.notReady("Torus wallet is not initialized");

              case 3:
                _context3.next = 5;
                return _get(_getPrototypeOf(TorusWalletAdapter.prototype), "disconnect", this).call(this);

              case 5:
                _context3.next = 7;
                return this.torusInstance.logout();

              case 7:
                this.torusInstance.hideTorusButton();

                if (options.cleanup) {
                  this.status = ADAPTER_STATUS.NOT_READY;
                  this.torusInstance = null;
                } else {
                  // ready to be connected again
                  this.status = ADAPTER_STATUS.READY;
                }

                this.rehydrated = false;
                this.emit(ADAPTER_EVENTS.DISCONNECTED);

              case 11:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function disconnect() {
        return _disconnect.apply(this, arguments);
      }

      return disconnect;
    }()
  }, {
    key: "getUserInfo",
    value: function () {
      var _getUserInfo = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee4() {
        var userInfo;
        return _regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                if (!(this.status !== ADAPTER_STATUS.CONNECTED)) {
                  _context4.next = 2;
                  break;
                }

                throw WalletLoginError.notConnectedError("Not connected with wallet");

              case 2:
                if (this.torusInstance) {
                  _context4.next = 4;
                  break;
                }

                throw WalletInitializationError.notReady("Torus wallet is not initialized");

              case 4:
                _context4.next = 6;
                return this.torusInstance.getUserInfo("");

              case 6:
                userInfo = _context4.sent;
                return _context4.abrupt("return", userInfo);

              case 8:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getUserInfo() {
        return _getUserInfo.apply(this, arguments);
      }

      return getUserInfo;
    }()
  }, {
    key: "setAdapterSettings",
    value: function setAdapterSettings(options) {
      if (this.status === ADAPTER_STATUS.READY) return;

      if (options !== null && options !== void 0 && options.sessionTime) {
        this.sessionTime = options.sessionTime;
      }

      if (options !== null && options !== void 0 && options.clientId) {
        this.clientId = options.clientId;
      }
    }
  }]);

  return TorusWalletAdapter;
}(BaseEvmAdapter);

export { TorusWalletAdapter };
